core/src/main/java/com/novoda/noplayer/internal/exoplayer/NoPlayerAdsLoader.java

```diff
@@ -179,39 +182,54 @@ public void onTimelineChanged(Timeline timeline, @Nullable Object manifest, int
             return;
         }
 
-        // handle only pre-roll start on timeline change
-        if (reason == Player.TIMELINE_CHANGE_REASON_PREPARED) {
-            handleAdvertStart();
-        }
+        updateState();
     }
 
-    private void handleAdvertStart() {
-        if (advertHasNotStarted()) {
-            adGroupIndex = player.getCurrentAdGroupIndex();
-            adIndexInGroup = player.getCurrentAdIndexInAdGroup();
+    private void updateState() {
+        boolean wasPlayingAd = playingAdvert;
+        playingAdvert = player.isPlayingAd();
+
+        int previousAdvertGroupIndex = advertGroupIndex;
+        int previousAdvertIndexInAdvertGroup = advertIndexInAdvertGroup;
+
+        advertGroupIndex = playingAdvert ? player.getCurrentAdGroupIndex() : C.INDEX_UNSET;
+        advertIndexInAdvertGroup = playingAdvert ? player.getCurrentAdIndexInAdGroup() : C.INDEX_UNSET;
+        boolean advertFinished = wasPlayingAd && advertIndexInAdvertGroup != previousAdvertIndexInAdvertGroup;
 
-            if (canPlayAdverts(adGroupIndex)) {
-                notifyAdvertStart(advertBreaks.get(adGroupIndex));
-            } else {
-                resetAdvertPosition();
-            }
+        if (advertFinished) {
+            adPlaybackState = adPlaybackState.withPlayedAd(previousAdvertGroupIndex, previousAdvertIndexInAdvertGroup);
+            updateAdPlaybackState();
+            notifyAdvertEnd(previousAdvertGroupIndex, previousAdvertIndexInAdvertGroup);
+        }
+
+        if (playingAdvert && advertIndexInAdvertGroup != previousAdvertIndexInAdvertGroup) {
+            notifyAdvertStart(advertBreaks.get(advertGroupIndex));
```

Comment:
Same thing here about making it symmetrical to `notifyAdvertEnd`. We could pass in `advertGroupIndex` and `advertIndexInAdvertGroup` (even though they are fields)

Reply:
cdc57a7